Remove "NON-solutions" about adding object methods in Object.create (#16290)
mdn/content: files/en-us/web/javascript/reference/global_objects/object/create/index.md

If `proto` is neither of these a {{jsxref("TypeError")}} is thrown.

## Object with `null` prototype

A new object with `null` prototype can behave in unexpected ways, because it doesn't inherit any object methods from `Object.prototype`. This is especially true when debugging, since common object-property
converting/detecting utility functions may generate errors, or lose information
(especially if using silent error-traps that ignore errors). For example, here are two
objects:

```js

> oco = {};   // create a normal object

Reviewer: ```suggestion
const oco = {};   // create a normal object
```

> ocn = Object.create(null); // create a "null" object

Reviewer: ```suggestion
const ocn = Object.create(null); // create an object with "null" prototype
```

messages)_

```js

> ob = {}; ob.po = oco; ob.pn = ocn; // create a compound object using the test objects from above as property values

Reviewer: ```suggestion
const ob = {}; ob.po = oco; ob.pn = ocn; // create a compound object using the test objects from above as property values
```
would be greate if we put these on separate lines

> showProperties(ob) // display top-level properties
- po: [object Object]
- Error: Cannot convert object to primitive value
```

Note that only first property gets shown, due to `ob.po` being traversed first. But if the same object is created in a different order...

```js

> ob = {}; ob.pn = ocn; ob.po = oco; // create same compound object again, but create same properties in different order

Reviewer: ```suggestion
const ob = {}; ob.pn = ocn; ob.po = oco; // create same compound object again, but create same properties in different order
```

```

Note that neither property gets shown.

Such a different order may arise statically via disparate fixed codings such
as here, but also dynamically via whatever the order any such property-adding
code-branches actually get executed at runtime as depends on inputs and/or
random-variables.

Beware, also, that using [`Object.entries()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries) on an object created via `Object.create()`
will result in an empty array being returned, because the first argument is the _prototype_, not the object's own properties.

```js
const obj = Object.create({ a: 1, b: 2 });

> console.log(Object.entries(obj)); // shows "[]"
```

### Adding object methods back

As demonstrated above, lack of default object methods can make debugging unwieldy. We can add the `toString` method back to the "null" object by simply assigning it one:

```js
ocn = Object.create(null); // create "null" object (same as before)

ocn.toString = Object.prototype.toString; // since new object lacks toString, add the original generic one back

> ocn.toString() // shows "[object Object]"
> "ocn is: " + ocn // shows "ocn is: [object Object]"

> ob = {}; ob.pn = ocn; ob.po = oco; // create a compound object (same as before)

Reviewer: ```suggestion
const ob = {}; ob.pn = ocn; ob.po = oco; // create a compound object (same as before)
```

> ShowProperties(ob) // display top-level properties
- po: [object Object]
- pn: [object Object]
```

Different from normal objects, `toString` here is an own property of `ocn`, instead of being on its prototype. This is because, well, `ocn` has no (`null`) prototype.

Resetting the new object's prototype works even better:

```js

> ocn = Object.create(null);                  // create "null" object (same as before)

Reviewer: ```suggestion
const ocn = Object.create(null);                  // create "null" object (same as before)
```

// create same compound object again, but create same properties in different order
const obj = {};
obj.pn = nullProtoObj;
obj.po = normalObj;

> ShowProperties( ob ) // display top-level properties
- Error: Cannot convert object to primitive value
showProperties(obj); // display top-level properties
// - Error: Cannot convert object to primitive value
```

Note that neither property gets shown.
```

Note that such a different order may arise statically via disparate fixed codings such
Such a different order may arise statically via disparate fixed codings such
as here, but also dynamically via whatever the order any such property-adding
code-branches actually get executed at runtime as depends on inputs and/or
random-variables. Then again, the actual iteration order is not guaranteed no matter
what the order members are added.
random-variables.

Be aware of, also, that using Object.entries() on an object created via Object.create()
will result in an empty array being returned.

```js
var obj = Object.create({ a: 1, b: 2 });

> console.log(Object.entries(obj)); // shows "[]"

> ```

Author: I've also removed this section because there's, really, nothing special about it. It's just that the first argument is the prototype, not own properties. Putting it here makes people think there's something inherently different from object literal syntax. Let me know if I should move it to anywhere else.

messages)_

```js
ob={}; ob.po=oco; ob.pn=ocn; // create a compound object using the test objects from above as property values
ob = {}; ob.po = oco; ob.pn = ocn; // create a compound object using the test objects from above as property values

> ShowProperties( ob ) // display top-level properties
> showProperties(ob) // display top-level properties
- po: [object Object]
- Error: Cannot convert object to primitive value

Note that only first property gets shown.
```

_(But if the same object is created in a different order -- at least in some

> implementations...)_

Author: I removed the notion that the iteration order is implementation-specific. AFAICT, it's well-defined in the latest ES spec. Couldn't find decisive words in the spec (could just be my reading problems), but seems like it from [this SO answer](https://stackoverflow.com/questions/30076219/does-es6-introduce-a-well-defined-order-of-enumeration-for-object-properties)

> > ShowProperties(ob) // display top-level properties

Reviewer 2: ```suggestion
> showProperties(ob) // display top-level properties
```

random-variables. Then again, the actual iteration order is not guaranteed no matter
what the order members are added.

Be aware of, also, that using [`Object.entries()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries) on an object created via `Object.create()`
will result in an empty array being returned, because the first argument is the _prototype_, not the object's own properties.

```js
var obj = Object.create({ a: 1, b: 2 });

> console.log(Object.entries(obj)); // shows "[]"
```

## Adding back object methods

As demonstrated above, lack of default object methods can make debugging unwieldy. We can add the `toString` method back to the "null" object by simply assigning it one:

```js
ocn = Object.create(null); // create "null" object (same as before)

> ocn.toString = Object.prototype.toString; // since new object lacks toString, add the original generic one back

Author: `ocn.toString = toString` is a bit hacky. It's effectively `ocn.toString = globalThis.toString`. Using `Object.prototype.toString` makes it more explicit.
